src/RulesDeploy.ts

```diff
@@ -53,72 +58,96 @@ export class RulesDeploy {
    * Compile all rulesets tied to this deploy, rejecting on first
    * compilation error.
    */
-  compile(): Promise<any> {
-    const promises: any[] = [];
-    _.forEach(this.rulesFiles, (files: RulesetFile[], filename: string) => {
-      promises.push(this._compileRuleset(filename, files));
-    });
-    return Promise.all(promises);
+  async compile(): Promise<void> {
+    for (const filename of Object.keys(this.rulesFiles)) {
+      const files = this.rulesFiles[filename];
+      await this._compileRuleset(filename, files);
+    }
   }
 
   /**
    * Create rulesets for each file added to this deploy, and record
    * the name for use in the release process later.
+   *
+   * If the ruleset to create is identical to the latest existing ruleset,
+   * then we record the existing ruleset name instead of creating a duplicate.
+   *
+   * @param {string} service The service to create a ruleset.
+   *    Valid values are "cloud.firestore" and "firebase.storage".
+   * @return All the names of the rulesets that were created.
    */
-  createRulesets(): Promise<any> {
-    const promises: any[] = [];
-    _.forEach(this.rulesFiles, (files: RulesetFile[], filename: any) => {
-      utils.logBullet(
-        clc.bold.cyan(this.type + ":") + " uploading rules " + clc.bold(filename) + "..."
-      );
-      promises.push(
-        gcp.rules.createRuleset(this.options.project, files).then((rulesetName: any) => {
-          this.rulesetNames[filename] = rulesetName;
-        })
-      );
-    });
-    return Promise.all(promises).catch(async (err) => {
-      if (err.status === QUOTA_EXCEEDED_STATUS_CODE) {
+  async createRulesets(service: string): Promise<string[]> {
+    const createdRulesetNames: string[] = [];
+
+    const latestRulesetName = await gcp.rules.getLatestRulesetName(this.options.project, service);
+    let latestRulesetContent: RulesetFile[] | undefined;
+    if (latestRulesetName) {
+      latestRulesetContent = await gcp.rules.getRulesetContent(latestRulesetName);
+    }
+
+    try {
+      for (const filename of Object.keys(this.rulesFiles)) {
+        const files = this.rulesFiles[filename];
+        if (_.isEqual(files, latestRulesetContent)) {
```

Comment:
can we extract this de-duplication logic into a separate function?

maybe
```
async createRulesets(service: Service): Promise<RulesetName[]> {
  const changedFiles = findChangedFiles(this.files, await fetchCurrentRules(service));
  ...
}
```
?

Reply:
I took some of the logic above away, but I think decoupling the decisions here from the logging and uploading makes it more complex than it's worth. (I'd still have to loop over everything again or return multiple sets, or be removing entire chunks of logic into helpers which won't really help much anyway.)